zhusuan/distributions/multivariate.py

```diff
@@ -29,6 +33,158 @@
+                       ' mean. Expected', tf.convert_to_tensor(expected_shape),
+                       'got', tf.convert_to_tensor(actual_shape)]
+            assert_ops = [tf.assert_equal(x, y, err_msg)
+                          for x, y in zip(expected_shape, actual_shape)]
+        else:
+            expected_shape = tf.concat(
+                    [tf.shape(self._mean), [self._n_dim]], axis=0)
+            actual_shape = tf.shape(self._cov_tril)
+            msg = ['MVNTriL.cov_tril should have compatible shape with '
+                   'mean. Expected', expected_shape, ' got ', actual_shape]
+            assert_ops = [tf.assert_equal(expected_shape, actual_shape, msg)]
+
+        with tf.control_dependencies(assert_ops):
+            self._cov_tril = tf.identity(self._cov_tril)
+
+        dtype = assert_same_float_dtype([(self._mean, 'MVNTriL.mean'),
+                                         (self._cov_tril, 'MVNTriL.cov_tril')])
+        super(MultivariateNormalTriL, self).__init__(
+            dtype=dtype,
+            param_dtype=dtype,
+            is_continuous=True,
+            is_reparameterized=is_reparameterized,
+            use_path_derivative=use_path_derivative,
+            group_ndims=group_ndims,
+            **kwargs)
+
+    @property
+    def mean(self):
+        """The mean of the normal distribution."""
+        return self._mean
+
+    @property
+    def cov_tril(self):
+        """
+        The lower triangular matrix in the cholosky decomposition of the
+        covariance.
+        """
+        return self._cov_tril
+
+    def _value_shape(self):
+        return tf.convert_to_tensor([self._n_dim], tf.int32)
+
+    def _get_value_shape(self):
+        if isinstance(self._n_dim, int):
+            return tf.TensorShape([self._n_dim])
+        return tf.TensorShape([None])
+
+    def _batch_shape(self):
+        return tf.shape(self.mean)[:-1]
+
+    def _get_batch_shape(self):
+        if self.mean.get_shape():
+            return self.mean.get_shape()[:-1]
+        return tf.TensorShape(None)
+
+    def _sample(self, n_samples):
+        mean, cov_tril = self.mean, self.cov_tril
+        if not self.is_reparameterized:
+            mean = tf.stop_gradient(mean)
+            cov_tril = tf.stop_gradient(cov_tril)
+
+        def tile(t):
+            new_shape = tf.concat([[n_samples], tf.ones_like(tf.shape(t))], 0)
+            return tf.tile(tf.expand_dims(t, 0), new_shape)
+
+        batch_mean = tile(mean)
+        batch_cov = tile(cov_tril)
+        # n_dim -> n_dim x 1 for matmul
+        batch_mean = tf.expand_dims(batch_mean, -1)
+        noise = tf.random_normal(tf.shape(batch_mean), dtype=self.dtype)
+        samples = tf.matmul(batch_cov, noise) + batch_mean
+        samples = tf.squeeze(samples, -1)
+        # Update static shape
+        static_n_samples = n_samples if isinstance(n_samples, int) else None
+        if self.get_batch_shape():
```

Comment:
Seems no need for this `if`? concatenate will work correctly when the shape is not defined.

Reply:
`n_samples` could be a tensor, and the constructor will fail

Reply:
I mean this line: `if self.get_batch_shape()`?